### PR TITLE
feat(internal): auto-generate snippets

### PR DESCRIPTION
**Pull request by tbpg**
I created a new package to make it easier to run than calling the binary. I included a binary for local development.

I needed to install `build-base` in the Docker image to get `gcc`, which was needed to `Load` all packages.

I modified `gapics.go` to replace `genproto` in every `go.mod` because the changes need to be visible everywhere to `Load` every module.

I ran the generator locally and it seems to work. I only included one directory in this PR because there are a lot of files. I can upload the rest in this PR after initial review.

Updates #3916.

### CODE REVIEW

File: internal/gapicgen/generator/gapics.go [hunk 1]
```diff
@@ -18,11 +18,14 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"io/fs"
```
Comment (codyoss) [inline internal/gapicgen/generator/gapics.go:21]: We should update the go version in the go.mod to 1.16 if this package is now being used.

File: internal/gapicgen/generator/gapics.go [hunk 2]
```diff
@@ -86,25 +100,85 @@ func (g *GapicGenerator) Regen(ctx context.Context) error {
...
+		return nil
+	})
+}
+
+func goModTidy(dir string) error {
+	log.Printf("[%s] running go mod tidy\n", dir)
+	c := command("go", "mod", "tidy")
+	c.Dir = dir
+	c.Env = []string{
+		fmt.Sprintf("PATH=%s", os.Getenv("PATH")), // TODO(deklerk): Why do we need to do this? Doesn't seem to be necessary in other exec.Commands.
```
Comment (codyoss) [inline internal/gapicgen/generator/gapics.go:134]: In a future PR I want to look into removing these, it is not obvious to me why they are needed either...

File: internal/gapicgen/generator/gapics.go [hunk 3]
```diff
@@ -86,25 +100,85 @@ func (g *GapicGenerator) Regen(ctx context.Context) error {
...
+			if err := fn(filepath.Dir(path)); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func goModTidy(dir string) error {
+	log.Printf("[%s] running go mod tidy\n", dir)
```
Comment (codyoss) [inline internal/gapicgen/generator/gapics.go:130]: Do you want the extra newline? One should be appended already with the log package.

File: internal/gapicgen/generator/gapics.go [hunk 4]
```diff
@@ -86,25 +100,85 @@ func (g *GapicGenerator) Regen(ctx context.Context) error {
...
+	if err := forEachMod(g.googleCloudDir, g.dropModReplaceGenproto); err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// forEachMod runs the given function with the directory of
+// every non-internal module.
+func forEachMod(rootDir string, fn func(dir string) error) error {
```
Comment (codyoss) [inline internal/gapicgen/generator/gapics.go:112]: Great idea!

File: internal/gapicgen/generator/gapics.go [hunk 5]
```diff
@@ -86,25 +100,85 @@ func (g *GapicGenerator) Regen(ctx context.Context) error {
...
+	})
+}
+
 // addModReplaceGenproto adds a genproto replace statement that points genproto
 // to the local copy. This is necessary since the remote genproto may not have
 // changes that are necessary for the in-flight regen.
-func (g *GapicGenerator) addModReplaceGenproto() error {
-	log.Println("adding temporary genproto replace statement")
+func (g *GapicGenerator) addModReplaceGenproto(dir string) error {
+	log.Printf("[%s] adding temporary genproto replace statement\n", dir)
```
Comment (codyoss) [inline internal/gapicgen/generator/gapics.go:175]: nit: same as above about newline.

File: internal/gapicgen/generator/gapics.go [hunk 6]
```diff
@@ -115,15 +189,14 @@ go mod edit -replace "google.golang.org/genproto@$GENPROTO_VERSION=$GENPROTO_DIR
 
 // dropModReplaceGenproto drops the genproto replace statement. It is intended
 // to be run after addModReplaceGenproto.
-func (g *GapicGenerator) dropModReplaceGenproto() error {
-	log.Println("removing genproto replace statement")
+func (g *GapicGenerator) dropModReplaceGenproto(dir string) error {
+	log.Printf("[%s] removing genproto replace statement\n", dir)
```
Comment (codyoss) [inline internal/gapicgen/generator/gapics.go:193]: nit: same as above

**Review by codyoss — COMMENTED** [review]
Just a couple of small things or else looks good. Thanks for some of the cleanup you did as well!

**Review by noahdietz — COMMENTED** [review]
LGTM ill let Cody approve

File: internal/gapicgen/generator/gapics.go (hunk 1 above)
Comment (tbpg) [reply internal/gapicgen/generator/gapics.go:21]: Done.

File: internal/gapicgen/generator/gapics.go (hunk 3 above)
Comment (tbpg) [reply internal/gapicgen/generator/gapics.go]: Done.

File: internal/gapicgen/generator/gapics.go (hunk 2 above)
Comment (tbpg) [reply internal/gapicgen/generator/gapics.go:134]: Ack.

File: internal/gapicgen/generator/gapics.go (hunk 5 above)
Comment (tbpg) [reply internal/gapicgen/generator/gapics.go]: Done.

File: internal/gapicgen/generator/gapics.go (hunk 6 above)
Comment (tbpg) [reply internal/gapicgen/generator/gapics.go]: Done.

File: internal/gapicgen/generator/gapics.go (hunk 4 above)
Comment (tbpg) [reply internal/gapicgen/generator/gapics.go:112]: Thanks! :smile:

**Review by tbpg — COMMENTED** [review]
Thanks!

**Review by codyoss — APPROVED** [review]
LGTM

**Pull request merged**